Expand/collapse icon

anastasia opened the pull request:
Displaying an expand or collapse icon in links list.
Expand icon is only displayed on link item hover or tab into list. 
Collapse icon is displayed when details have been expanded.

![vtdgsq6rgs](https://user-images.githubusercontent.com/1494102/37617562-de8b498a-2b89-11e8-97c5-9be6396cd4b3.gif)


- tested in chrome and firefox. 
- unable to get past the search bar in safari -- merits further investigation

rebeccacremona commented on perma_web/perma/templates/user_management/create-link.html:
```diff
@@ -170,10 +169,24 @@ <h3 class="link-header body-ah">Your Perma Links</h3>
-
           <div class="row item-row row-no-bleed _isDraggable" data-link_id="{{ guid }}">
+            <!-- this input is here only so on details expand or collapse
+            the user doesn't wind up all the way up at the top of the screen again -->
```
Hooray for explicitly handling focus! This is indeed super important. 

When are you finding that focus gets reset to the body element (e.g., the "top of the page")? That most commonly results when a) the element that had keyboard focus is zapped from the DOM by JS, or b) when you try to follow a link to an in-page anchor that doesn't exist. For instance, `<a href="#no-element-on-the-page-has-this-id">` or c) when you follow a link whose href is "#" and the js that runs doesn't include `event.preventDefault()`. (In this last case, it's usually a good hint that you are really dealing with a "button" rather than a "link", semantically.

I think what you want to have happen here is have focus toggle between the "Show Link Details" button/link and the "Expand Link Details" button/link, kind of like when you are toggling a dropdown open/closed. For that to work, each of these `<a>` needs a unique id, but since you have `{{ guid }}` sounds like that won't be too tough!

rebeccacremona commented on perma_web/static/js/links-list.module.js:
```diff
@@ -1,17 +1,17 @@
-var LinkHelpers = require('./helpers/link.helpers.js');
-var HandlebarsHelpers = require('./helpers/handlebars.helpers.js');
-var APIModule = require('./helpers/api.module.js');
+let DOMHelpers = require('./helpers/dom.helpers.js');
```
thanks for all these `let`s!

anastasia commented on perma_web/static/bundles/create.js:
```diff
@@ -283,14 +283,14 @@ webpackJsonp([1],[
 	  var requestCount = 20,
 	      requestData = { limit: requestCount, offset: 0 },
-	      endpoint;
+	      endpoint = void 0;
```
This looks like it's added in the compiled file

rebeccacremona replied:
Oh whoops.

rebeccacremona commented on perma_web/perma/templates/user_management/create-link.html:
```diff
@@ -170,10 +169,24 @@ <h3 class="link-header body-ah">Your Perma Links</h3>
           <div class="row item-row row-no-bleed _isDraggable" data-link_id="{{ guid }}">
+            <!-- this input is here only so on details expand or collapse
+            the user doesn't wind up all the way up at the top of the screen again -->
+            <input style="display: none;" name="{{ guid }}">
```
Update on above: looks like you've already got JS that does this without IDs. So never mind that particular part.

anastasia commented:
Seriously?? Man! That's good to hear. I've been unable to get into the links list section at all in Safari. It goes from foldertree to search bar, and then goes all the way up to the top. I wonder if it's some kind of weird setting on my end, or maybe I need to update the browser.

anastasia commented on perma_web/perma/templates/user_management/create-link.html:
```diff
@@ -170,10 +169,24 @@ <h3 class="link-header body-ah">Your Perma Links</h3>
+                <!-- expand arrow -->
+                <a href="#{{ guid }}"
+                   class="_visuallyHidden toggle-details expand-details"
+                   title="Show Link Details">
```
Ack! I definitely had an aria-label earlier in this pull request, and then removed it when I simplified the html structure. Adding it back in.

anastasia commented on perma_web/static/bundles/create.js:
```diff
@@ -340,6 +340,12 @@ webpackJsonp([1],[
 	  var template = HandlebarsHelpers.renderTemplate(templateId, templateArgs);
 	  linkTable.append(template);
+	  var toggleDetailsIcon = $('.toggle-details');
+	  $(toggleDetailsIcon).keydown(function (e) {
```
Yes, button is best. It simplifies _so much code_

rebeccacremona commented:
This is looking so good!! It's so close: I just found one tiny bug. When I expand link details and try to click into the input fields to change the title, etc., the details collapse again. (Keyboard works though :-)) That one last thing, and we're good to go! Thanks for all your work on this.

anastasia commented:
Ack, you're right! Moved the click event to another element and added one small stylistic change (arrow was appearing too far to the left on small screens)
